### ISSUE TITLE
`GPU.Input` type do not work for CPU mode

### ISSUE THREAD

**Issue by anoopelias** [gpujs/gpu.js#243]
Problem: Passing `GPU.Input` type do not work for CPU mode
Version: 1.0.0 (From npm)

Test Code:
```
import GPU, { input } from 'gpu.js';

function run() {
  const gpu = new GPU({
    mode: 'cpu'
  });

  const opt = { output: [3, 3] };
  const kernel = gpu.createKernel(function(values) {
    return values[this.thread.x] + values[this.thread.y];
  }, opt);


  const gpuInput = input(new Float32Array([2, 4, 6, 1, 3, 5, 8, 10, 12]), [3, 3]);
  const output = kernel(gpuInput);
  console.log(output);
}

run();
```
Output
```
(3) [Array(3), Array(3), Array(3)]
0:(3) [NaN, NaN, NaN]
1:(3) [NaN, NaN, NaN]
2:(3) [NaN, NaN, NaN]
length:3
__proto__:Array(0)
```
Changing mode to `gpu` and then getting proper output:
```
(3) [Array(3), Array(3), Array(3)]
0:(3) [4, 6, 8]
1:(3) [6, 8, 10]
2:(3) [8, 10, 12]
length:3
__proto__:Array(0)
```

**Comment by anoopelias** [comment]
There is a custom implementation for `Input` type in webgl-kernel.js [here](https://github.com/gpujs/gpu.js/blob/4173dd9610b9c04fa064e7e2e12b5f066d7eeac7/src/backend/web-gl/kernel.js#L627), but it is missing in cpu-kernel, see [here](https://github.com/gpujs/gpu.js/blob/4173dd9610b9c04fa064e7e2e12b5f066d7eeac7/src/backend/cpu/kernel.js#L42).

Is this simply not done yet? Or am I missing something?

**Comment by robertleeplummerjr** [comment]
That is correct, it just has not yet been implemented.  The key difference on GPU over CPU is that at this time on GPU the Array's are _always_ flat, but there is some architecture in place to help flatten them, which is expensive, but isn't worth making everyone have to pre-flatten their arrays. `GPU.input` skips this step, allowing for very fast computation without having to flatten, because the data _is_ pre-flattened.  I imagine the best means of which to implement this would be to use input, and then on the CPUFunctionBuilder we'd have an option (or detection?) to simply use them, altering the compiled kernel.  So: `input1[this.thread.z][this.thread.y][this.thread.x]` would become something more like: `input1[this.thread.hypotheticalndexOrSomething]`.

**Comment by robertleeplummerjr** [comment]
Addressing this now, fyi.

**Comment by robertleeplummerjr** [comment]
Solved!

**Comment by robertleeplummerjr** [comment]
How this was solved: the parser actually rewrites the javascript so it creates a single index for the xyz coordinate, and rather than add `value[this.thread.z][this.thread.y][this.thread.x]` it tracks the state it is writing, and writes: `value[this + math + here]`


This was a fun one!
